naming/dns_resolver_test.go

```diff
@@ -0,0 +1,304 @@
+		Op:       op,
+		Addr:     addr,
+		Metadata: AddrMetadataGRPCLB{AddrType: GRPCLB, ServerName: lb},
+	}
+}
+
+func newSRVRR(target string, port uint16) *net.SRV {
+	return &net.SRV{
+		Target: target,
+		Port:   port,
+	}
+}
+
+var updateTc = []struct {
+	oldAddrs []string
+	newAddrs []string
+	want     []*Update
+}{
+	{
+		[]string{},
+		[]string{"1.0.0.1"},
+		[]*Update{newUpdate(Add, "1.0.0.1")},
+	},
+	{
+		[]string{"1.0.0.1"},
+		[]string{"1.0.0.1"},
+		[]*Update{},
+	},
+	{
+		[]string{"1.0.0.0"},
+		[]string{"1.0.0.1"},
+		[]*Update{newUpdate(Delete, "1.0.0.0"), newUpdate(Add, "1.0.0.1")},
+	},
+	{
+		[]string{"1.0.0.1"},
+		[]string{"1.0.0.0"},
+		[]*Update{newUpdate(Add, "1.0.0.0"), newUpdate(Delete, "1.0.0.1")},
+	},
+	{
+		[]string{"1.0.0.1"},
+		[]string{"1.0.0.1", "1.0.0.2", "1.0.0.3"},
+		[]*Update{newUpdate(Add, "1.0.0.2"), newUpdate(Add, "1.0.0.3")},
+	},
+	{
+		[]string{"1.0.0.1", "1.0.0.2", "1.0.0.3"},
+		[]string{"1.0.0.0"},
+		[]*Update{newUpdate(Add, "1.0.0.0"), newUpdate(Delete, "1.0.0.1"), newUpdate(Delete, "1.0.0.2"), newUpdate(Delete, "1.0.0.3")},
+	},
+	{
+		[]string{"1.0.0.1", "1.0.0.3", "1.0.0.5"},
+		[]string{"1.0.0.2", "1.0.0.3", "1.0.0.6"},
+		[]*Update{newUpdate(Delete, "1.0.0.1"), newUpdate(Add, "1.0.0.2"), newUpdate(Delete, "1.0.0.5"), newUpdate(Add, "1.0.0.6")},
+	},
+}
+
+func converToMap(u []*Update) map[string]*Update {
+	m := make(map[string]*Update)
+	for _, v := range u {
+		m[v.Addr] = v
+	}
+	return m
+}
+
+func TestCompileUpdate(t *testing.T) {
+	var w dnsWatcher
+	for _, c := range updateTc {
+		w.curAddrs = make([]*Update, len(c.oldAddrs))
+		newUpdates := make([]*Update, len(c.newAddrs))
+		for i, a := range c.oldAddrs {
+			w.curAddrs[i] = &Update{Addr: a}
+		}
+		for i, a := range c.newAddrs {
+			newUpdates[i] = &Update{Addr: a}
+		}
+		r := w.compileUpdate(newUpdates)
+		if !reflect.DeepEqual(converToMap(c.want), converToMap(r)) {
+			t.Errorf("w(%+v).compileUpdate(%+v) = %+v, want %+v", c.oldAddrs, c.newAddrs, updatesToSlice(r), updatesToSlice(c.want))
+		}
+	}
+}
+
+var validAddrTc = []struct {
+	addr string
+	want error
+}{
+	// TODO(yuxuanli): More false cases?
+	{"www.google.com", nil},
+	{"foo.bar:12345", nil},
+	{"127.0.0.1", nil},
+	{"127.0.0.1:12345", nil},
+	{"[::1]:80", nil},
+	{"[2001:db8:a0b:12f0::1]:21", nil},
+	{":80", nil},
+	{"127.0.0...1:12345", nil},
+	{"[fe80::1%lo0]:80", nil},
+	{"golang.org:http", nil},
+	{"[2001:db8::1]:http", nil},
+	{":", nil},
+	{"", errMissingAddr},
+	{"[2001:db8:a0b:12f0::1", fmt.Errorf("invalid target address %v", "[2001:db8:a0b:12f0::1")},
+}
+
+func TestResolveFunc(t *testing.T) {
+	r, err := NewDNSResolver()
+	if err != nil {
+		t.Errorf("%v", err)
+	}
+	for _, v := range validAddrTc {
+		_, err := r.Resolve(v.addr)
+		if !reflect.DeepEqual(err, v.want) {
+			t.Errorf("Resolve(%s) = %v, want %v", v.addr, err, v.want)
+		}
+	}
+}
+
+//TODO(yuxuanli): Do we need to test with net.LookupHost, net.LookupSRV and real target? If not, delete this.
```

Comment:
No, fake only seems fine to me.  That would require integration testing to bring up a DNS server, presumably.